mmseg/models/decode_heads/protoseg_head.py

```diff
@@ -0,0 +1,156 @@
+                ModuleHelper.BNReLU(dim_in, bn_type=bn_type),
+                nn.Conv2d(dim_in, proj_dim, kernel_size=1)
+            )
+
+    def forward(self, x):
+        return F.normalize(self.proj(x), p=2, dim=1)
+
+class Protoseg_head(nn.Module):
+    def __init__(self,in_channels):
+        super(Hrnet,self).__init__()
+        self.in_channel=in_channels
+        self.num_prototype=10
+        self.update_prototype=True
+        self.gamma=0.999
+        self.pretrain_prototype=False
+        self.num_classes=19
+        self.cls_head = nn.Sequential(
+            nn.Conv2d(in_channels, in_channels, kernel_size=3, stride=1, padding=1),
+            ModuleHelper.BNReLU(in_channels, bn_type="torchbn"),
+            nn.Dropout2d(0.10)
+        )
+        
+        self.prototypes = nn.Parameter(torch.zeros(self.num_classes, self.num_prototype, in_channels),
+                                       requires_grad=True)
+
+        self.proj_head = ProjectionHead(in_channels, in_channels)
+        self.feat_norm = nn.LayerNorm(in_channels)
+        self.mask_norm = nn.LayerNorm(self.num_classes)
+    def prototype_learning(self, _c, out_seg, gt_seg, masks):
+        pred_seg = torch.max(out_seg, 1)[1]
+        mask = (gt_seg == pred_seg.view(-1))
+        cosine_similarity = torch.mm(_c, self.prototypes.view(-1, self.prototypes.shape[-1]).t())
+        proto_logits = cosine_similarity
+        proto_target = gt_seg.clone().float()
+
+        protos = self.prototypes.data.clone()
+        for k in range(self.num_classes):
+            init_q = masks[..., k]
+            init_q = init_q[gt_seg == k, ...]
+            if init_q.shape[0] == 0:
+                continue
+
+            q, indexs = distributed_sinkhorn(init_q)
+
+            m_k = mask[gt_seg == k]
+
+            c_k = _c[gt_seg == k, ...]
+
+            m_k_tile = repeat(m_k, 'n -> n tile', tile=self.num_prototype)
+
+            m_q = q * m_k_tile
+
+            c_k_tile = repeat(m_k, 'n -> n tile', tile=c_k.shape[-1])
+
+            c_q = c_k * c_k_tile  # n x embedding_dim
+
+            f = m_q.transpose(0, 1) @ c_q  # self.num_prototype x embedding_dim
+
+            n = torch.sum(m_q, dim=0)
+
+            if torch.sum(n) > 0 and self.update_prototype is True:
+                f = F.normalize(f, p=2, dim=-1)
+
+                new_value = momentum_update(old_value=protos[k, n != 0, :], new_value=f[n != 0, :],
+                                                momentum=self.gamma, debug=False)
+                protos[k, n != 0, :] = new_value
+
+            proto_target[gt_seg == k] = indexs.float() + (self.num_prototype * k)
+
+        self.prototypes = nn.Parameter(l2_normalize(protos),
+                                           requires_grad=False)
+
+        if dist.is_available() and dist.is_initialized():
+            protos = self.prototypes.data.clone()
+            dist.all_reduce(protos.div_(dist.get_world_size()))
```

Comment:
Can't find the definition of `dist`.

Reply:
MMEngine implements all the distributed training logits, there is no need to implement it in the model.

Reply:
> Can't find the definition of `dist`.
This fix